src/http/gateway/resources/gateway.js

```diff
@@ -0,0 +1,141 @@
+'use strict'
+
+const debug = require('debug')
+const log = debug('jsipfs:http-gateway')
+log.error = debug('jsipfs:http-gateway:error')
+const pull = require('pull-stream')
+const toPull = require('stream-to-pull-stream')
+const fileType = require('file-type')
+const mime = require('mime-types')
+const Stream = require('readable-stream')
+
+const GatewayResolver = require('../resolver')
+const PathUtils = require('../utils/path')
+
+module.exports = {
+  checkCID: (request, reply) => {
+    if (!request.params.cid) {
+      return reply({
+        Message: 'Path Resolve error: path must contain at least one component',
+        Code: 0
+      }).code(400).takeover()
+    }
+
+    return reply({
+      ref: `/ipfs/${request.params.cid}`
+    })
+  },
+  handler: (request, reply) => {
+    const ref = request.pre.args.ref
+    const ipfs = request.server.app.ipfs
+
+    return GatewayResolver
+           .resolveMultihash(ipfs, ref)
+           .then((data) => {
+             ipfs
+             .files
+             .cat(data.multihash)
+             .then((stream) => {
+               if (ref.endsWith('/')) {
+                  // remove trailing slash for files
+                 return reply
+                        .redirect(PathUtils.removeTrailingSlash(ref))
+                        .permanent(true)
+               } else {
+                 if (!stream._read) {
+                   stream._read = () => {}
+                   stream._readableState = {}
+                 }
+                 //  response.continue()
+                 let filetypeChecked = false
+                 let stream2 = new Stream.PassThrough({highWaterMark: 1})
+                 let response = reply(stream2).hold()
+
+                 pull(
+                   toPull.source(stream),
+                   pull.drain((chunk) => {
```

Comment:
This means we are buffering the whole response in memory before responding to the client